internal/pkg/gateway/api.go

```diff
@@ -273,6 +270,44 @@ func (gs *Server) Endorse(ctx context.Context, request *gp.EndorseRequest) (*gp.
 	return endorseResponse, nil
 }
 
+// responseStatus unpacks the proposal response and error values that are returned from ProcessProposal and
+// determines how the gateway should react (retry?, close connection?).
+// Uses the grpc canonical status error codes and their recommended actions.
+// Returns:
+// - response successful (bool)
+// - error message extracted from the err or generated from 500 proposal response
+// - should the gateway retry (only the Evaluate() uses this)
+// - should the gateway close the connection and remove the peer from its registry
+func (gs *Server) responseStatus(response *peer.ProposalResponse, err error) (bool, string, bool, bool) {
+	// if the ProcessProposal function returned an err, then it must have been from the unpackProposal() or preProcess() stages
+	// preProcess does all the signature and ACL checking. In either case, no point retrying, or closing the connection (it's a client error)
+	if err != nil {
+		if response == nil {
+			// if there is no ProposalResponse, then this must have been generated by grpc in response to an unavailable peer
+			// - close the connection and retry on another
+			return false, err.Error(), true, true
+		}
+		return false, err.Error(), false, false
+	}
+	if response.Response.Status < 200 || response.Response.Status >= 400 {
+		// if response contains a payload, then it must have been generated by the peer
```

Comment:
```suggestion
		// if response contains a payload, then it must have been generated by the peer
```
```suggestion
		// if response doesn't contain a payload, then it must have been generated by the peer rather than by chaincode
```

Reply:
yes, these comments could be better. I've reworded them and repositioned them.